x/fdsmr/node_test.go

```diff
@@ -0,0 +1,380 @@
+		},
+	}
+	expiry := time.Now().Add(time.Hour).Unix()
+	txs := []dsmrtest.Tx{
+		{
+			ID:      ids.GenerateTestID(),
+			Expiry:  expiry,
+			Sponsor: codec.EmptyAddress,
+		},
+	}
+
+	n := New[testDSMR, dsmrtest.Tx](
+		testDSMR{
+			AcceptF: func(_ context.Context, _ dsmr.Block) (dsmr.ExecutedBlock[dsmrtest.Tx], error) {
+				return dsmr.ExecutedBlock[dsmrtest.Tx]{
+					BlockHeader: dsmr.BlockHeader{},
+					ID:          ids.ID{},
+					Chunks: []dsmr.Chunk[dsmrtest.Tx]{
+						{
+							UnsignedChunk: dsmr.UnsignedChunk[dsmrtest.Tx]{
+								Producer:    ids.NodeID{},
+								Beneficiary: codec.Address{},
+								Expiry:      0,
+								Txs:         txs,
+							},
+							Signer:    [48]byte{},
+							Signature: [96]byte{},
+						},
+					},
+				}, nil
+			},
+		},
+		b,
+	)
+
+	r.NoError(n.BuildChunk(
+		context.Background(),
+		txs,
+		expiry,
+		codec.EmptyAddress,
+	))
+	r.Len(b.pending[codec.EmptyAddress], 1)
+
+	_, err := n.Accept(context.Background(), dsmr.Block{})
+	r.NoError(err)
+	r.Empty(b.pending[codec.EmptyAddress])
+}
+
+// Tests that txs are un-bonded after expiry
+func TestUnbondOnExpiry(t *testing.T) {
+	r := require.New(t)
+
+	b := testBonder{
+		pending: map[codec.Address]chan struct{}{
+			codec.EmptyAddress: make(chan struct{}, 1),
+		},
+	}
+
+	expiry := time.Now().Add(time.Second).Unix()
+	txs := []dsmrtest.Tx{
+		{
+			ID:      ids.GenerateTestID(),
+			Expiry:  expiry,
+			Sponsor: codec.EmptyAddress,
+		},
+	}
+
+	n := New[testDSMR, dsmrtest.Tx](testDSMR{}, b)
+	r.NoError(n.BuildChunk(
+		context.Background(),
+		txs,
+		expiry,
+		codec.EmptyAddress,
+	))
+	// block until the tx is un-bonded
+	<-b.pending[codec.EmptyAddress]
+}
+
+type testDSMR struct {
+	BuildChunkF func(
+		ctx context.Context,
+		txs []dsmrtest.Tx,
+		expiry int64,
+		beneficiary codec.Address,
+	) error
+	AcceptF func(
+		ctx context.Context,
+		block dsmr.Block,
+	) (dsmr.ExecutedBlock[dsmrtest.Tx], error)
+}
+
+func (t testDSMR) BuildChunk(ctx context.Context, txs []dsmrtest.Tx, expiry int64, beneficiary codec.Address) error {
+	if t.BuildChunkF == nil {
+		return nil
+	}
+
+	return t.BuildChunkF(ctx, txs, expiry, beneficiary)
+}
+
+func (testDSMR) BuildBlock(dsmr.Block, int64) (dsmr.Block, error) {
+	panic("implement me")
+}
+
+func (testDSMR) Verify(context.Context, dsmr.Block, dsmr.Block) error {
+	panic("implement me")
+}
+
+func (t testDSMR) Accept(ctx context.Context, block dsmr.Block) (dsmr.ExecutedBlock[dsmrtest.Tx], error) {
+	if t.AcceptF == nil {
+		return dsmr.ExecutedBlock[dsmrtest.Tx]{}, nil
+	}
+
+	return t.AcceptF(ctx, block)
+}
+
+type testBonder struct {
+	pending map[codec.Address]chan struct{}
+}
+
+func (b testBonder) Bond(tx dsmrtest.Tx) bool {
+	pending, ok := b.pending[tx.GetSponsor()]
+	if !ok {
+		return false
+	}
+
+	if len(pending) == cap(pending) {
+		return false
+	}
+
+	pending <- struct{}{}
+	return true
+}
+
+func (b testBonder) Unbond(tx dsmrtest.Tx) {
+	<-b.pending[tx.GetSponsor()]
+}
```

Comment:
Can we simplify this if we switch away from using goroutines to mark txs as expired and unbond txs when the chain time advances past their expiry?

Reply:
Yeah we can just use a map.